pod2daemon/csidriver/driver/driver.go

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) 2022 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package driver
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+	"strings"
+
+	csi "github.com/container-storage-interface/spec/lib/go/csi"
+	log "github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+)
+
+const (
+	VER            string = "0.1"
+	CONFIG_FILE    string = "/etc/calico/csi/nodeagent.json"
+	NODEAGENT_HOME string = "/var/run/nodeagent"
+	MOUNT_DIR      string = "/mount"
+	CREDS_DIR      string = "/creds"
+	LOG_LEVEL_WARN string = "WARNING"
+	CSI_SOCK       string = "/csi/csi.sock"
+)
+
+// ConfigurationOptions may be used to setup the driver.
+// These are optional and most users will not depened on them and will instead use the defaults.
+type ConfigurationOptions struct {
+	// Location on the node's filesystem where the driver will host the
+	// per workload directory and the credentials for the workload.
+	// Default: /var/run/nodeagent
+	NodeAgentManagementHomeDir string `json:"nodeagent_management_home,omitempty"`
```

Comment:
do we want to use k8s style json tags here?

e.g., `nodeagentManagementHomeDir`?

I'm assuming this config will normally be passed in via a configmap if needed?

Reply:
> I'm assuming this config will normally be passed in via a configmap if needed?

A lot of this code I borrowed directly from the flexvolume so if anyone had a special config for their flexvolume, their migration would be pretty seamless. I haven't included an easy way to make this happen with our manifests. Anyone who knows that they would need to set the configmap as a volume to the pod and configure the configmap appropriately so that it writes all this data to `/etc/calico/csi/nodeagent.json` in the pod could make it happen, but my intention for now is to keep it all locked down (except for `logLevel` and `csiSockPath`, which are passed in as arguments). Theoretically a user could also want to change `nodeagentManagementHomeDir` but given how our instructions look at https://projectcalico.docs.tigera.io/security/app-layer-policy, it is highly unlikely anyone will want to (or has) changed it.

Ideally, I would rather do away with reading the entire config from a file or configmap and just send the few individual knobs we do have (e.g. socket path, nodeagent directory, log level) in to the driver as arguments.